example-application/libraries/message-queue-client.js

```diff
@@ -69,17 +128,68 @@ class MessageQueueClient extends EventEmitter {
       //Not awaiting because some MQ client implementation get back to fetch messages again only after handling a message
       onMessageCallback(theNewMessage.content.toString())
         .then(() => {
+          this.emit('ack', theNewMessage);
           this.channel.ack(theNewMessage);
         })
         .catch((error) => {
-          this.channel.nack(theNewMessage);
+          this.channel.nack(theNewMessage, false, this.requeue);
+          this.emit('nack', theNewMessage);
           error.isTrusted = true; //Since it's related to a single message, there is no reason to let the process crash
-          errorHandler.handleError(error);
+          //errorHandler.handleError(error);
         });
     });
 
     return;
   }
+
+  setRequeue(newValue) {
+    this.requeue = newValue;
+  }
+
+  // This function stores all the MQ events in a local data structure so later
+  // one query this
+  countEvents() {
+    const eventsToListen = ['nack', 'ack', 'publish'];
+    if (this.eventsRecorder !== undefined) {
+      return; // Already initialized and set up
+    }
+    this.eventsRecorder = {};
+    eventsToListen.forEach((eventToListenTo) => {
+      this.eventsRecorder[eventToListenTo] = {
+        count: 0,
+        lastEventData: null,
+        name: eventToListenTo,
+      };
+      this.on(eventToListenTo, (eventData) => {
+        this.eventsRecorder[eventToListenTo].count++;
+        this.eventsRecorder[eventToListenTo].lastEventData = eventData;
+        this.emit('message-queue-event', {
+          name: eventToListenTo,
+          eventsRecorder: this.eventsRecorder,
+        });
+      });
+    });
+  }
+
+  resolveIfEventExceededThreshold(eventName, threshold, resolve) {
+    if (this.eventsRecorder[eventName].count >= threshold) {
+      resolve({
+        name: eventName,
+        lastEventData: this.eventsRecorder[eventName].lastEventData,
+        count: this.eventsRecorder[eventName].count,
+      });
+    }
+  }
+  // Helper methods for testing - Resolves/fires when some event happens
+  async waitFor(eventName, howMuch) {
+    return new Promise((resolve, reject) => {
+      // The first resolve is for cases where the caller has approached AFTER the event has already happen
+      this.resolveIfEventExceededThreshold(eventName, howMuch, resolve);
+      this.on('message-queue-event', (eventInfo) => {
```

Comment:
Great function 👍🏻
One small thing is that I suggest you clean the handler when resolved called as you can reach the default max listeners by just adding tests